sources/host-ctr/cmd/host-ctr/main.go

```diff
@@ -575,16 +579,80 @@ func cleanUp(containerdSocket string, namespace string, containerID string) erro
 	return nil
 }
 
+// parseImageURISpecialRegions mimics the parsing in ecr.ParseImageURI but constructs the canonical ECR references while
+// skipping certain checks. We only do this for special regions that are not yet supported by the aws-go-sdk.
+// Referenced source: https://github.com/awslabs/amazon-ecr-containerd-resolver/blob/a5058cf091f4fc573813a032db37a9820952f1f9/ecr/ref.go#L70-L71
+func parseImageURISpecialRegions(input string) (string, error) {
+	ecrRefPrefixMapping := map[string]string{
+		"il-central-1": "ecr.aws/arn:aws:ecr:il-central-1:",
+	}
+	// Matching on account, region
+	matches := ecrRegex.FindStringSubmatch(input)
+	if len(matches) < 3 {
+		return "", fmt.Errorf("invalid image URI: %s", input)
+	}
+	account := matches[1]
+	region := matches[2]
+
+	// Need to include the full repository path and the imageID (e.g. /eks/image-name:tag)
+	tokens := strings.SplitN(input, "/", 2)
+	if len(tokens) != 2 {
+		return "", fmt.Errorf("invalid image URI: %s", input)
+	}
+	fullRepoPath := tokens[len(tokens)-1]
+	// Run simple checks on the provided repository.
+	switch {
+	case
+		// Must not be empty
+		fullRepoPath == "",
+		// Must not have a partial/unsupplied label
+		strings.HasSuffix(fullRepoPath, ":"),
+		// Must not have a partial/unsupplied digest specifier
+		strings.HasSuffix(fullRepoPath, "@"):
+		return "", errors.New("incomplete reference provided")
+	}
+
+	// Parse out image reference's to validate.
+	ref, err := reference.Parse("repository/" + fullRepoPath)
+	if err != nil {
+		return "", err
+	}
+	// If the digest is provided, check that it is valid.
+	if ref.Digest() != "" {
+		err := ref.Digest().Validate()
+		// Digest may not be supported by the client despite it passing against
+		// a rudimentary check. The error is different in the passing case, so
+		// that's considered a passing check for unavailable digesters.
+		//
+		// https://github.com/opencontainers/go-digest/blob/ea51bea511f75cfa3ef6098cc253c5c3609b037a/digest.go#L110-L115
+		if err != nil && err != digest.ErrDigestUnsupported {
+			return "", fmt.Errorf("%v: %w", "invalid image URI", err)
+		}
+	}
+
+	// Get the ECR image reference prefix from the AWS region
+	ecrRefPrefix := ecrRefPrefixMapping[region]
+	return ecrRefPrefix + account + ":repository/" + fullRepoPath, nil
```

Comment:
The first test case I wrote fails where I expected it to just work:

```
❯ go test ./...
--- FAIL: TestEcrSpecialRegionsParser (0.00s)
    --- FAIL: TestEcrSpecialRegionsParser/Parse_amazon-ecr-resolver_region (0.00s)
        main_test.go:46:
                Error Trace:    /home/ubuntu/workspace/bottlerocket-os/bottlerocket/sources/host-ctr/cmd/host-ctr/main_test.go:46
                Error:          Not equal:
                                expected: "ecr.aws/arn:aws:ecr:us-west-2:111111111111:repository/bottlerocket/container:1.2.3"
                                actual  : "111111111111:repository/bottlerocket/container:1.2.3"

                                Diff:
                                --- Expected
                                +++ Actual
                                @@ -1 +1 @@
                                -ecr.aws/arn:aws:ecr:us-west-2:111111111111:repository/bottlerocket/container:1.2.3
                                +111111111111:repository/bottlerocket/container:1.2.3
                Test:           TestEcrSpecialRegionsParser/Parse_amazon-ecr-resolver_region
FAIL
FAIL    host-ctr/cmd/host-ctr   0.016s
FAIL
```

It looks like with a `us-west-2` region, which is bundled by `aws-sdk-go` and should be handled by the `amazon-ecr-resolver`, falls back to using our `ecrRegPrefixMapping` which does not have the `us-west-2` region key and returns an empty string, resulting in `111111111111:repository/bottlerocket/container:1.2.3`.

Is this the expected behavior?

Reply:
We probably need to check the map access here?

```go
	val, ok := ecrRefPrefixMapping[]
	if !ok {
                return "", fmt.Errorf("%v: %w", "invalid region in internal mapping", region)
	}
```

Otherwise we will always be masking unknown values in the prefix mapping by assuming an empty string.

Here's a little go playground to demonstrate this as well https://go.dev/play/p/W_8XzLuWOwp

Reply:
This also makes me wonder what the actual expected behavior for completely invalid regions is:

```go
                {
                        "Fallback to public ECR on invalid region",
                        "111111111111.dkr.ecr.outer-space.amazonaws.com/bottlerocket/container:1.2.3",
                        nil,
                        "111111111111:repository/bottlerocket/container:1.2.3",
                },
```

So, for region `outer-space` in this tase case, what's the expected result? It looks like the same thing is happening where the value from `ecrRefPrefixMapping` (which doesn't exist) is returning the empty string and is not populating the `ecrRefPrefix` portion of the returned image uri

Reply:
We can probably disregard the error I'm seeing with this function on `us-west-2` since calling the `amazon-ecr-resolver` _shouldn't_ error out when called on those regions _before_ we make it to the host-ctr's `parseImageURISpecialRegions`

https://github.com/etungsten/bottlerocket/blob/cc6c6530311ec5be4d96bfe72c04688bf6f791de/sources/host-ctr/cmd/host-ctr/main.go#L642-L651

Although that makes me abit nervous since there's a clear failure path where a valid region is returned in error by the ecr-resolver and is not covered by our subsequent function call.

Reply:
> It looks like with a us-west-2 region, which is bundled by aws-sdk-go and should be handled by the amazon-ecr-resolver, falls back to using our ecrRegPrefixMapping which does not have the us-west-2 region key and returns an empty string, resulting in 111111111111:repository/bottlerocket/container:1.2.3.
Is this the expected behavior?

Yes, that's expected behavior. The fallback only triggers when `ecr.ParseImageURI` fails to parse out the ECR image source.


> So, for region outer-space in this test case, what's the expected result? It looks like the same thing is happening where the value from ecrRefPrefixMapping (which doesn't exist) is returning the empty string and is not populating the ecrRefPrefix portion of the returned image uri

We error since we're limiting the fallback to only `il-central-1` and no other "special" region that isn't yet supported by the `aws-go-sdk`. This is only a temporary workaround for `il-central-1` that is meant to removed when a more stable and long-term fix is available.

> Although that makes me abit nervous since there's a clear failure path where a valid region is returned in error by the ecr-resolver and is not covered by our subsequent function call.

If `ecr.ParseImageURI` errors for whatever reason on a ECR image URL belonging to some "valid" region (valid meaning supported by the `aws-go-sdk`. We will fail regardless in the fallback, so the end result is the same.